### PR TITLE
Adds Bloodsuckers

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

Ports Bloodsuckers, including the unmerged Revenge Vassal I was working on.

Possibly outdated (I haven't read over it since I made it) hackMD here about it: https://hackmd.io/@fulpstation/BkEvrMTBK

## How does it improve JollyStation

I don't want Jolly to get $20

Bloodsuckers is a conversion antag that is more complicated for the people expected to know more (the bloodsuckers, who have their own tools to learn what to do, and time before they do anything, as well as being opt-in through their prefs since they aren't converted from anything), and simple for the people who are forced into the position (the converted Vassals), making it ease nicely unlike Blood cult, while not being as simple and barebones as Revs.

Bloodsuckers have different paths they can go but it typically revolves around your own lair and the allies you get whom you have to rely on, because they have a minute of weakness every 10 minutes, they are obligated to get some form of help if they don't want to be killed, so cooperation between team antagonist allies are required.

They have very limited ways of getting blood so they have to manage to find and secure source of it, because stuff like Iron, nanites (rip), cursed heart, ect, don't restore Bloodsucker blood, which I find generally leads to funny interactions of trying to coerce someone into giving you blood, or pulling off a heist from Medical to steal their bloodbags.

Bloodsuckers also follow a Masquerade code, enforced by eachother. If a Bloodsucker breaks the Masquerade, their HUD changes, Vassals can be stolen, and all other Bloodsuckers have to kill them, as a reference to VtM where Bloodhunts were called on people who were a liability of revealing the nature of Bloodsuckers to humanity. No one is supposed to know that Bloodsuckers exist, except the Curator.

For non-Bloodsuckers, it also affects the round:
- Curators can catch Bloodsuckers who don't disguise themselves well enough
- Vassals (converted people) can also be upgraded, if the Bloodsucker wants. The Favorite Vassal is more loyal to the Bloodsucker than any other, gets more abilities, or becomes a fully new Bloodsucker in the case of Ventrue (one of the clans)
- The second type of Vassal is the Revenge Vassal. If the Bloodsucker dies, the revenge vassal has to take care of all the ex-vassals by bringing them back into the fold and keeping them addicted to artificial bloodsucker blood, basically keeping the Bloodsucker's legacy alive, even if they will never come back.

Conversion is also optional if you're mindshielded or an antag, and since they don't roundremove like Cultists, there's generally more freedom to do what you want, rather than play to 'win' necessarily.

## Changelog

:cl: TheSwain, JohnFulpWillard
add: Bloodsuckers
/:cl:

### CODE REVIEW

File: code/__DEFINES/role_preferences.dm [hunk 1]
```diff
@@ -99,6 +99,11 @@
 #define ROLE_BATTLECRUISER_CAPTAIN "Battlecruiser Captain"
 #define ROLE_VENUSHUMANTRAP "Venus Human Trap"
 
+// jollystation edits
+#define ROLE_BLOODSUCKER "Bloodsucker"
+#define ROLE_VAMPIRICACCIDENT "Vampiric Accident"
+#define ROLE_BLOODSUCKERBREAKOUT "Bloodsucker Breakout"
+// jollystation edit ends
```
Comment (Jolly-66) [inline code/__DEFINES/role_preferences.dm:106]:
```suggestion
//  NON-MODULAR CHANGES: Bloodsuckers
#define ROLE_BLOODSUCKER "Bloodsucker"
#define ROLE_VAMPIRICACCIDENT "Vampiric Accident"
#define ROLE_BLOODSUCKERBREAKOUT "Bloodsucker Breakout"
// NON-MODULAR CHANGES END
```

File: code/__DEFINES/role_preferences.dm [hunk 2]
```diff
@@ -145,6 +150,12 @@ GLOBAL_LIST_INIT(special_roles, list(
 	// I'm not too sure why these are here, but they're not moving.
 	ROLE_PAI = 0,
 	ROLE_SENTIENCE = 0,
+
+	// jollystation edit
+	ROLE_BLOODSUCKER = 0,
+	ROLE_VAMPIRICACCIDENT = 0,
+	ROLE_BLOODSUCKERBREAKOUT = 0,
+	// jollystation edit ends
```
Comment (Jolly-66) [inline code/__DEFINES/role_preferences.dm:158]:
```suggestion
//  NON-MODULAR CHANGES: Blood suckers
	ROLE_BLOODSUCKER = 0,
	ROLE_VAMPIRICACCIDENT = 0,
	ROLE_BLOODSUCKERBREAKOUT = 0,
	//  NON-MODULAR CHANGES END

**Review by Jolly-66 — COMMENTED** [review]
Just dropping these for now.

I would suggest looking at our pain subsystem and see what you can add from that into this.

**Review by Jolly-66 — CHANGES_REQUESTED** [review]
Dropping this here since I had a bit to sleep on it.

I would like to see some implementations of pain here.
Be it for the Bloodsuckers themselves, vassals, or the unfortunate individuals.
Ideally, it would be nice to see it implemented for all three.

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -16,12 +16,22 @@
...
-talestation_modules/code/__DEFINES/pain_defines.dm @prodirus
-talestation_modules/code/datums/pain/pain_helpers.dm @prodirus
-talestation_modules/code/datums/pain/pain.dm @prodirus
-talestation_modules/code/modules/mob/living/carbon/pain/ @prodirus
-talestation_modules/code/modules/reagents/chemistry/reagents/pain_reagents.dm @prodirus
+/talestation_modules/code/__DEFINES/pain_defines.dm @prodirus
+/talestation_modules/code/datums/pain/pain_helpers.dm @prodirus
+/talestation_modules/code/datums/pain/pain.dm @prodirus
+/talestation_modules/code/modules/mob/living/carbon/pain/ @prodirus
+/talestation_modules/code/modules/reagents/chemistry/reagents/pain_reagents.dm @prodirus
```
Comment (Jolly-66) [inline .github/CODEOWNERS:27]: Could you atomize these out? You can, in theory, still leave it in the PR, but I'm asking if you can make a separate PR for just the codeowners tidying.

Comment (JohnFulpWillard) [reply .github/CODEOWNERS:27]: It's just adding a / at the stat so they work

**Comment by Jolly-66** [comment]
Prodirus will be reviewing this when they get time.

File: talestation_modules/code/modules/antagonists/bloodsuckers/powers/vassal/distress.dm [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+/datum/action/bloodsucker/distress/ActivatePower(trigger_flags)
+	. = ..()
+	var/turf/open/floor/target_area = get_area(owner)
+	var/datum/antagonist/vassal/vassaldatum = owner.mind.has_antag_datum(/datum/antagonist/vassal)
+
+	owner.balloon_alert(owner, "you call out for your master!")
+	to_chat(vassaldatum.master.owner, "<span class='userdanger'>[owner], your loyal Vassal, is desperately calling for aid at [target_area]!</span>")
+
+	var/mob/living/carbon/user = owner
+	user.adjustBruteLoss(10)
```
Comment (ZephyrTFA) [inline talestation_modules/code/modules/antagonists/bloodsuckers/powers/vassal/distress.dm:22]: should probably let the user know its causes damage

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm [hunk 5]
```diff
@@ -0,0 +1,429 @@
+/// Runs from COMSIG_LIVING_LIFE, handles Bloodsucker constant proccesses.
+/datum/antagonist/bloodsucker/proc/LifeTick()
```
Comment (ZephyrTFA) [inline talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm:2]: PascalCase

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+	description = "<span class='boldwarning'>I have been scorched by the unforgiving rays of the sun.</span>\n"
+	mood_change = -6
+	timeout = 6 MINUTES
+
+///Candelabrum's mood event to non Bloodsucker/Vassals
+/datum/mood_event/vampcandle
+	description = "<span class='boldwarning'>Something is making your mind feel... loose.</span>\n"
+	mood_change = -15
+	timeout = 5 MINUTES
+
```
Comment (ZephyrTFA) [inline talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm:429]: A lot of the procs in this file are using PascalCase

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: talestation_modules/code/__DEFINES/antag_defines.dm [hunk 7]
```diff
@@ -131,3 +131,135 @@
...
+
+/**
+ * Traits
+ */
+/// Your heart doesn't beat
+#define TRAIT_NOPULSE "nopulse"
+/// Falsifies Health analyzer blood levels
+#define TRAIT_MASQUERADE "masquerade"
+/// Your body is literal room temperature. Does not make you immune to the temp
+#define TRAIT_COLDBLOODED "coldblooded"
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/__DEFINES/antag_defines.dm:154]: XXX_TRAIT is for source traits, these aren't sources.

File: talestation_modules/code/__DEFINES/antag_defines.dm [hunk 8]
```diff
@@ -131,3 +131,135 @@
...
+#define BP_CANT_USE_WHILE_UNCONSCIOUS (1<<4)
+
+/// This Power can be purchased by Bloodsuckers
+#define BLOODSUCKER_CAN_BUY (1<<0)
+/// This is a Default Power that all Bloodsuckers get.
+#define BLOODSUCKER_DEFAULT_POWER (1<<1)
+/// This Power can be purchased by Tremere Bloodsuckers
+#define TREMERE_CAN_BUY (1<<2)
+/// This Power can be purchased by Vassals
+#define VASSAL_CAN_BUY (1<<3)
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/__DEFINES/antag_defines.dm:236]: You see it often, but it's not a requirement or standard. flags_1, foods, mechs, species changesource flags, and many others are just examples of ones that don't follow that.

File: talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/vassal/vassal_datum.dm [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	owner.enslave_mind_to_creator(master.owner.current)
+	owner.current.log_message("has been vassalized by [master.owner.current]!", LOG_ATTACK, color="#960000")
+	/// Give Recuperate Power
+	BuyPower(new /datum/action/bloodsucker/recuperate)
+	/// Give Objectives
+	var/datum/objective/bloodsucker/vassal/vassal_objective = new
+	vassal_objective.owner = owner
+	objectives += vassal_objective
+	/// Give Vampire Language & Hud
+	owner.current.grant_all_languages(FALSE, FALSE, TRUE)
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/vassal/vassal_datum.dm:98]: That just gives them omnitongue, so they can speak any language they already know. It's the alternative solution to a TG edit because tongues use a static list.

File: talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/vassal/vassal_datum.dm [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+		to_chat(master.owner, span_cultbold("You feel the bond with your vassal [owner.current] has somehow been broken!"))
+
+/// If we weren't created by a bloodsucker, then we cannot be a vassal (assigned from antag panel)
+/datum/antagonist/vassal/can_be_owned(datum/mind/new_owner)
+	if(!master)
+		return FALSE
+	return ..()
+
+/// When a Bloodsucker gets FinalDeath, all Vassals are freed - This is a Bloodsucker proc, not a Vassal one.
+/datum/antagonist/bloodsucker/proc/FreeAllVassals()
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/vassal/vassal_datum.dm:165]: cause it's old

File: talestation_modules/code/modules/antagonists/bloodsuckers/rulesets/bloodsucker_rulesets.dm [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+	if(!is_special_character(target))
+		return FALSE
+	/// Checks if the person is an antag banned from being vassalized, stored in bloodsucker's datum.
+	for(var/datum/antagonist/antag_datum in target.mind.antag_datums)
+		if(antag_datum.type in vassal_banned_antags)
+			//message_admins("DEBUG VASSAL: Found Invalid: [antag_datum] // [antag_datum.type]")
+			return TRUE
+//	message_admins("DEBUG VASSAL: Valid Antags! (total of [target.antag_datums.len])")
+	// WHEN YOU DELETE THE ABOVE: Remove the 3 second timer on converting the vassal too.
+	return FALSE
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/rulesets/bloodsucker_rulesets.dm:214]: some antagonists cant be vassalized, its ``vassal_banned_antags`` on bloodsucker's datum.

File: talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/bloodsucker_datum.dm [hunk 12]
```diff
@@ -0,0 +1,768 @@
...
+
+/// End Sol, if you're the last Bloodsucker
+/datum/team/vampireclan/proc/check_cancel_sunlight()
+	// No minds in the clan? Delete Sol.
+	if(members.len <= 1)
+		message_admins("Sol has been deleted due to the lack of Bloodsuckers")
+		QDEL_NULL(bloodsucker_sunlight)
+
+/// Buying powers
+/datum/antagonist/bloodsucker/proc/BuyPower(datum/action/bloodsucker/power)
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/bloodsucker_datum.dm:428]: I kept this because it's an important proc and this case makes it stand out.

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_daylight.dm [hunk 13]
```diff
@@ -0,0 +1,160 @@
+/// 1 minute
+#define TIME_BLOODSUCKER_DAY 60
+/// 10 minutes
+#define TIME_BLOODSUCKER_NIGHT 600
+/// 1.5 minutes
+#define TIME_BLOODSUCKER_DAY_WARN 90
+/// 30 seconds
+#define TIME_BLOODSUCKER_DAY_FINAL_WARN 30
+/// 5 seconds
+#define TIME_BLOODSUCKER_BURN_INTERVAL 5
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_daylight.dm:10]: It's there to undef it at the end, but I actually forgot to do so.

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_integration.dm [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+	if(!viewer.mind.has_antag_datum((/datum/antagonist/bloodsucker)) && !(viewer in bloodsuckerdatum.vassals))
+		if(!(HAS_TRAIT(viewer, TRAIT_BLOODSUCKER_HUNTER) && bloodsuckerdatum.broke_masquerade))
+			return FALSE
+	// Default String
+	var/returnString = "\[<span class='warning'><EM>[bloodsuckerdatum.ReturnFullName()]</EM></span>\]"
+	var/returnIcon = "[icon2html('talestation_modules/icons/bloodsuckers/vampiric.dmi', world, "bloodsucker")]"
+
+	// In Disguise (Veil)?
+	//if (name_override != null)
+	//	returnString += "<span class='suicide'> ([real_name] in disguise!) </span>"
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_integration.dm:96]: It hasn't been implemented yet.

File: talestation_modules/code/modules/antagonists/bloodsuckers/powers/vassal/distress.dm (hunk 4 above)
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/powers/vassal/distress.dm:22]: ``desc = "Injure yourself,``

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm (hunk 5 above)
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm:2]: this is an important proc so I want it to stand out.

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm (hunk 6 above)
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm:429]: changed the unimportant ones, kept the important ones,

File: talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/bloodsucker_hud.dm [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+/datum/antagonist/bloodsucker/proc/update_sunlight(value, amDay = FALSE)
+	if(!sunlight_display)
+		return
+	var/valuecolor
+	var/sunlight_display_icon = "sunlight_"
+	if(amDay)
+		sunlight_display_icon += "day"
+		valuecolor = "#FF5555"
+	else
+		switch(round(value, 1))
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/antagonists/bloodsucker_hud.dm:60]: this is a switch case, it's needed for the hud to be accurate

File: talestation_modules/code/__HELPERS/damage_helpers.dm [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+/mob/living/proc/getFireLoss_nonProsthetic()
+	return getFireLoss()
+
+/mob/living/carbon/getFireLoss_nonProsthetic()
+	var/amount = 0
+	for(var/obj/item/bodypart/chosen_bodypart as anything in bodyparts)
+		if(!IS_ORGANIC_LIMB(chosen_bodypart))
+			continue
+		amount += chosen_bodypart.burn_dam
+	return amount
```
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/__HELPERS/damage_helpers.dm:23]: at least for now, it's consistent with TG

File: talestation_modules/code/__DEFINES/antag_defines.dm (hunk 7 above)
[the remark this answers is not in the document]
Comment (JohnFulpWillard) [reply talestation_modules/code/__DEFINES/antag_defines.dm:154]: Unless you mean in what it's defining, in which case that is also wrong
image

File: talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm (hunk 6 above)
Comment (JohnFulpWillard) [reply talestation_modules/code/modules/antagonists/bloodsuckers/bloodsucker/bloodsucker_life.dm:429]: I also, in some ocassions, kept PascalCase as a way to keep track of things I've touched, and parts that is still from Swain's era, though I wasn't consistent with that and lost track of it.

**Comment by Jolly-66** [comment]
Throwing this up as a Test Merge until we get a population consensus on this. It'll remain like this until numbers are stable.

**Comment by JohnFulpWillard** [comment]
i dont want to maintain this sorry
